### ISSUE TITLE
Output chunks change when build stays the same

### ISSUE THREAD

**Issue by chriskrogh** [webpack/webpack#18837]
# Bug report

The original issue was reported here https://github.com/vercel/next.js/issues/70754, but I have a strong suspicion that the issue is with webpack and not with nextjs.

**What is the current behavior?**

when none of the source files change, webpack creates new output chunks for some of the pages.

<img width="1139" alt="Screenshot 2024-10-03 at 11 41 29 AM" src="https://github.com/user-attachments/assets/11cc1161-67b3-42e8-9902-6068ffc4b9b0">

https://www.diffchecker.com/OidQJCx0/

according to [the webpack docs](https://webpack.js.org/guides/caching/):
> The [contenthash] substitution will add a unique hash based on the content of an asset. When the asset's content changes, [contenthash] will also change.

if you compare the changed `page-[contenthash].js` chunks across builds, you'll notice that the contents are nearly identical, except for the webpack chunk IDs at the top of the file. the contents of the array are the same, it's just that the order can change across builds

<img width="1330" alt="Screenshot 2024-10-03 at 12 05 47 PM" src="https://github.com/user-attachments/assets/ca167455-eb34-4550-b63f-10ae00d3fba0">

https://www.diffchecker.com/Zp1tnr7X/

**i.e. because the chunk IDs aren't sorted at the top of the chunk, webpack treats these as "new" chunks, even though the actual code hasn't changed.**

**If the current behavior is a bug, please provide the steps to reproduce.**

https://github.com/chriskrogh/build-manifest-test

1. clone the repo and run `npm i`
2. run the nextjs build (`npm run build`) and take note of the `app-build-manifest.json`
3. run the nextjs build again and compare the `app-build-manifest.json`s across the two builds 

* you may want to sort them by route if they're not in the same order across builds
* you may also need to repeat step 3 if you observe identical chunk names. see above for why that is

**What is the expected behavior?**

* chunks should be the same when the build doesn't change

**Proposed solution**

* sort the webpack chunk ID array before writing to the file so that the array stays the same if the contents are the same

I'm willing to open a PR if someone points me to the right place.

**Other relevant information:**
webpack version:  5.90.0
Node.js version: 20.12.0
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.6.0: Mon Jul 29 21:14:30 PDT 2024; root:xnu-10063.141.2~1/RELEASE_ARM64_T6000
  Available memory (MB): 32768
  Available CPU cores: 10

**Comment by alexander-akait** [comment]
Maybe related too https://github.com/webpack/webpack/issues/17757

**Comment by chriskrogh** [comment]
@alexander-akait I'm pretty sure we just need to add a `.sort()` to the chunk ids before we write to the chunk. do you know where that logic might live in the codebase? I'm willing to test locally and put up a PR

**Comment by alexander-akait** [comment]
> I'm pretty sure we just need to add a .sort() to the chunk ids before we write to the chunk.

Before making changes we need to understand what and how they are created, I will try to figure it out in the near future

**Comment by gergokee** [comment]
i have same issue, and yes a simple sort() would solve it, i tested it

**Comment by alexander-akait** [comment]
I found the root of the problem and it is inside `nextjs` - https://github.com/webpack/webpack/blob/main/lib/optimize/FlagIncludedChunksPlugin.js#L50

`compilation.modules` is not deterministic and always has a different order of `modules` and due this `moduleBits` and `hash` are always different - https://github.com/webpack/webpack/blob/main/lib/optimize/FlagIncludedChunksPlugin.js#L65 and when comparing hashes - https://github.com/webpack/webpack/blob/main/lib/optimize/FlagIncludedChunksPlugin.js#L109 in some situations this condition is `true` or `false` - so the order is random. 

This can happen due to changing the order of modules in some async hooks. In pure webpack this does not happen because we do not do such things, but apparently in some `nextjs` plugin this happens. Unfortunately, debugging in which requires deeper knowledge in `nextjs`, this takes me much more time, I have already spent about 5-6 hours to find where the problem is and why. Anyway I intend to fix this because we have async hooks before this, which could potentially happen in other unofficial plugins too.
